packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+    ast::ClassDef::RHS_store exportedItems;
+};
+
+class NameFormatter {
+    const core::GlobalState &gs;
+
+public:
+    NameFormatter(const core::GlobalState &gs) : gs(gs) {}
+
+    void operator()(std::string *out, core::NameRef name) const {
+        out->append(name.data(gs)->shortName(gs));
+    }
+};
+
+// Gets the package name in `name` if applicable.
+PackageName getPackageName(core::MutableContext ctx, ast::TreePtr &name, bool errorOnInvalidName = false) {
+    PackageName pName;
+    pName.loc = name->loc;
+
+    auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(name);
+    while (constLit != nullptr) {
+        if (errorOnInvalidName && absl::StrContains(constLit->cnst.data(ctx)->shortName(ctx), "_")) {
+            // By forbidding package names to have an underscore, we can trivially convert between mangled names and
+            // unmangled names by replacing `_` with `::`.
+            if (auto e = ctx.beginError(constLit->loc, core::errors::Packager::InvalidPackageName)) {
+                e.setHeader("Package names cannot contain an underscore");
+                auto replacement = absl::StrReplaceAll(constLit->cnst.data(ctx)->shortName(ctx), {{"_", ""}});
+                auto nameLoc = constLit->loc;
+                // cnst is the last characters in the constant literal
+                nameLoc.beginLoc = nameLoc.endLoc - constLit->cnst.data(ctx)->shortName(ctx).size();
+
+                e.addAutocorrect(core::AutocorrectSuggestion{
+                    fmt::format("Replace `{}` with `{}`", constLit->cnst.data(ctx)->shortName(ctx), replacement),
+                    {core::AutocorrectSuggestion::Edit{core::Loc(ctx.file, nameLoc), replacement}}});
+            }
+        }
+        pName.fullName.emplace_back(constLit->cnst);
+        constLit = ast::cast_tree<ast::UnresolvedConstantLit>(constLit->scope);
+    }
+    reverse(pName.fullName.begin(), pName.fullName.end());
+    if (!pName.fullName.empty()) {
+        // Foo::Bar => Foo_Bar_Package
+        auto mangledName = absl::StrCat(absl::StrJoin(pName.fullName, "_", NameFormatter(ctx)), "_Package");
+        pName.mangledName = ctx.state.enterNameConstant(mangledName);
+    } else {
+        // Not a valid name.
+        pName.mangledName = core::NameRef::noName();
+    }
+    return pName;
+}
+
+bool isReferenceToPackageSpec(core::Context ctx, ast::TreePtr &expr) {
+    auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    return constLit != nullptr && constLit->cnst == core::Names::Constants::PackageSpec();
+}
+
+unique_ptr<ast::UnresolvedConstantLit> copyConstantLit(ast::UnresolvedConstantLit *lit) {
+    auto copy = lit->deepCopy();
+    // Cast from Expression to UnresolvedConstantLit.
+    auto copyUcl = ast::cast_tree<ast::UnresolvedConstantLit>(copy);
+    ENFORCE(copyUcl != nullptr);
+    unique_ptr<ast::UnresolvedConstantLit> rv(copyUcl);
+    copy.release();
+    return rv;
+}
```

Comment:
@elliottt is this allowed? I don't know that I've seen any `unique_ptr<ast::Expression>` that is not a `TreePtr` after the recent changes to `TreePtr`.

Reply:
The intermediate unique_ptr should be unnecessary. I would replace it with something like the following:

```cpp
unique_ptr<ast::UnresolvedConstantLit> copyConstantLit(ast::UnresolvedConstantLit *lit) {
    auto copy = lit->deepCopy();
    ENFORCE(ast::isa_tree<ast::UnresolvedConstantLit>(copy));
    return reinterpret_cast<ast::UnresolvedConstantLit>(copy.release());
}
```

Separately, why does it return a `unique_ptr`? is a `TreePtr` not useful in this case, or is it just more useful to have the `unique_pointer` because you know the type of thing that it holds?

Reply:
> is it just more useful to have the unique_pointer because you know the type of thing that it holds?

Yup. I can switch it to TreePtr if we're delurking all unique_ptrs to nodes; it just adds noise and makes the code harder to follow bc the reader needs to know the invariants w.r.t. what the TreePtr points to.

Reply:
That makes sense. I don't see any problem with this then 👍

Reply:
@elliottt 's code is returning the pointer that `copy` owns, so it's returning a freed resource -- so I think this code is fine as-is.

Reply:
Yes, my changes were relying on my incorrect assumption that there was an implicit constructor for unique_ptr available. It's good that there's not, and your original version looks good 👍